src/Compilers/Core/Portable/DiagnosticAnalyzer/AsyncQueue.cs

```diff
@@ -177,6 +177,21 @@ private bool CompleteCore()
             {
                 if (existingWaiters != null)
                 {
+                    // we have waiters and we have data.
+                    // allow some extra time to drain the queue before cancelling
+                    // but do not wait for too long.
+                    SpinWait.SpinUntil(() =>
+                    {
+                        lock (SyncObject)
+                        {
+                            return _data.Count == 0;
+                        }
+                    },
+                    10000);
+
+                    // cancel waiters.
+                    // NOTE: this could result in losing diagnostics
+                    //       see https://github.com/dotnet/roslyn/issues/11470
```

Comment:
AsyncQueue is not specifically for diagnostics. Please move these comments into the caller.